# HW_1_Python SciPy.Optimize

#Problem Statement: Solve the problem using Python SciPy.optimize
<br>
#Minimize: (x1-x2)^2+(x2+x3-2)^2+(x4-1)^2+(x5-1)^2
<br>
#Subject to:<br>
             x1+3x2 = 0 <br>
             x3+x4-2x5 = 0 <br>
             x2-x5 = 0 <br>
#Boundary Conditions : <br>
-10<= xi <= 10, i = 1,2,3,4,5 <br>
#Initial guesses of the solution <br>
[-1,-1,-1,-1,-1],[-10,-9,-8,-7,-6],[10,9,8,7,6],[10,10,10,10,10],[-10,-10,-10,-10,-10],[0,0,0,0,0]

# Code

In [141]:
# First Iteration with initial conditions [-1,-1,-1,-1,-1]
from scipy.optimize import minimize

# Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2 + (x[3]-1)**2 + (x[4]-1)**2

# Constraints
cons = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Boundary Conditions
bnds = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Execution
res = minimize(fun, (-1,-1,-1,-1,-1), method='SLSQP', bounds=bnds,constraints=cons)

# Output
res
print(res)

     fun: 4.093023255824878
     jac: array([-2.04651427, -0.18604708, -2.23256135, -2.23255301, -1.48837143])
 message: 'Optimization terminated successfully'
    nfev: 43
     nit: 7
    njev: 7
  status: 0
 success: True
       x: array([-0.76744284,  0.25581428,  0.62790506, -0.1162765 ,  0.25581428])


In [142]:
## Second Iteration with initial conditions [-10,-9,-8,-7,-6]
from scipy.optimize import minimize

# Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2+(x[3]-1)**2+(x[4]-1)**2

# Constraints
cons = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Boundary Conditions
bnds = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Execution
res = minimize(fun, (-10,-9,-8,-7,-6), method='SLSQP', bounds=bnds,constraints=cons)

# Output
res
print(res)

     fun: 4.093023255813954
     jac: array([-2.04651159, -0.18604654, -2.23255813, -2.23255813, -1.48837209])
 message: 'Optimization terminated successfully'
    nfev: 37
     nit: 6
    njev: 6
  status: 0
 success: True
       x: array([-0.76744186,  0.25581395,  0.62790697, -0.11627907,  0.25581395])


In [143]:
## Third Iteration with initial conditions [10,9,8,7,6]
from scipy.optimize import minimize

# Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2+(x[3]-1)**2+(x[4]-1)**2

# Constraints
cons = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Boundary Conditions
bnds = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Execution
res = minimize(fun, (10,9,8,7,6), method='SLSQP', bounds=bnds,constraints=cons)

# Output
res
print(res)

     fun: 4.093023255813954
     jac: array([-2.04651159, -0.18604648, -2.23255813, -2.23255813, -1.48837209])
 message: 'Optimization terminated successfully'
    nfev: 37
     nit: 6
    njev: 6
  status: 0
 success: True
       x: array([-0.76744185,  0.25581395,  0.62790697, -0.11627906,  0.25581395])


In [144]:
## Fourth Iteration with initial conditions [10,10,10,10,10]
from scipy.optimize import minimize

# Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2+(x[3]-1)**2+(x[4]-1)**2

# Constraints
cons = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Boundary Conditions
bnds = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Execution
res = minimize(fun, (10,10,10,10,10), method='SLSQP', bounds=bnds,constraints=cons)

# Output
res
print(res)

     fun: 4.093023258012402
     jac: array([-2.04654074, -0.18606842, -2.23260909, -2.23248535, -1.48836482])
 message: 'Optimization terminated successfully'
    nfev: 48
     nit: 8
    njev: 8
  status: 0
 success: True
       x: array([-0.76745276,  0.25581759,  0.62787785, -0.11624268,  0.25581759])


In [145]:
## Fifth Iteration with initial conditions [-10,-10,-10,-10,-10]
from scipy.optimize import minimize

# Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2+(x[3]-1)**2+(x[4]-1)**2

# Constraints
cons = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Boundary Conditions
bnds = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Execution
res = minimize(fun, (-10,-10,-10,-10,-10), method='SLSQP', bounds=bnds,constraints=cons)

# Output
res
print(res)

     fun: 4.09302325604787
     jac: array([-2.04650199, -0.18603969, -2.23254162, -2.23258191, -1.48837447])
 message: 'Optimization terminated successfully'
    nfev: 48
     nit: 8
    njev: 8
  status: 0
 success: True
       x: array([-0.76743824,  0.25581275,  0.62791644, -0.11629094,  0.25581275])


In [146]:
## Sixth Iteration with initial conditions [0,0,0,0,0]
from scipy.optimize import minimize

# Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2+(x[3]-1)**2+(x[4]-1)**2

# Constraints
cons = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Boundary Conditions
bnds = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Execution
res = minimize(fun, (0,0,0,0,0), method='SLSQP', bounds=bnds,constraints=cons)

# Output
res
print(res)

     fun: 4.093023255813954
     jac: array([-2.04651159, -0.18604648, -2.23255813, -2.23255813, -1.48837209])
 message: 'Optimization terminated successfully'
    nfev: 20
     nit: 3
    njev: 3
  status: 0
 success: True
       x: array([-0.76744186,  0.25581395,  0.62790698, -0.11627907,  0.25581395])


# Change your initial guess, do you find different solutions?

|Iteration       |	      |Function            |      |X values| <br>
|:---------------|        |:----------------   |      |:--------------------------------------------------------------------|  
|__-1,-1,-1,-1,-1|	      |4.09302325"5824878" |      |-0.7674'4284',  0.25581428,  0.62790506, -0.1162765 ,  0.25581428__ <br>
|-10,-9,-8,-7,-6 |        |4.093023255813954   |      |-0.76744186,  0.25581395,  0.62790697, -0.11627907,  0.25581395 <br>
|10,9,8,7,6	     |        |4.093023255813954   |      |-0.76744185,  0.25581395,  0.62790697, -0.11627906,  0.25581395 <br>
|10,10,10,10,10	 |        |4.093023258012402   |      |-0.76745276,  0.25581759,  0.62787785, -0.11624268,  0.25581759 <br>
|__-10,-10,-10,-10,-10|	  |4.09302325"604787"  |      |-0.7674'3824',  0.25581275,  0.62791644, -0.11629094,  0.25581275__ <br>
|0,0,0,0,0	     |        |4.093023255813954   |      |-0.76744186,  0.25581395,  0.62790698, -0.11627907,  0.25581395 <br>

- Based on above iterations, I observed that with change in initial conditions there is very minimum change in the solution.
For example, when we consider the 2 scenarios(1st and 5th), we can see that value of function changes at 9th decimal point and value of x1 changes from 5th decimal point, which is very minimum.
- For some scenarios the function value is almost same and there is slight change in x values.